src/ai/defaultai.cc

```diff
@@ -3341,70 +3344,219 @@ void DefaultAI::diplomacy_actions(const Time& gametime) {
-				if (!other_alone) {
+				if (!other_alone && g_verbose) {
 					other_team_score_str = format(" and team score %d", ots);
 				}
-				accept = ots > my_team_score;
+				const int32_t my_effective_ts = plan_to_leave ? 0 : my_team_score;
+				accept = ots > my_effective_ts;
+				priority = ots - my_effective_ts;
+				if (accept && plan_to_leave) {
+					// Replace plan to leave with accepting invite, which also leaves old team
+					plan_priority = 0;
+				}
 			}
 
-			game().send_player_diplomacy(pda.other,
-			                             (pda.action == Widelands::DiplomacyAction::kInvite ?
-                                          (accept ? Widelands::DiplomacyAction::kAcceptInvite :
-                                                    Widelands::DiplomacyAction::kRefuseInvite) :
-                                          (accept ? Widelands::DiplomacyAction::kAcceptJoin :
-                                                    Widelands::DiplomacyAction::kRefuseJoin)),
-			                             pda.sender);
-			verb_log_dbg_time(
-			   gametime, "AI Diplomacy: Player(%d)%s %s %s player (%d) with diploscore %d%s\n",
-			   static_cast<unsigned int>(pda.other), myts_s.c_str(), accept ? "accepts" : "denies",
-			   pda.action == Widelands::DiplomacyAction::kInvite ? "the invitation of" : "to join",
-			   static_cast<unsigned int>(pda.sender), diploscore, other_team_score_str.c_str());
+			accept = accept && (planned_action == kNoAction || plan_priority < priority);
 
-			if (accept) {
-				request_accepted = true;
+			if (!accept) {
+				verb_log_dbg_time(
+				   gametime,
+				   "AI Diplomacy: Player(%d)%s denies the %s of player (%d) with diploscore %d%s\n",
+				   static_cast<unsigned int>(pda.other), myts_s.c_str(),
+				   pda.action == Widelands::DiplomacyAction::kInvite ? "invitation" : "join request",
+				   static_cast<unsigned int>(pda.sender), diploscore, other_team_score_str.c_str());
+				game().send_player_diplomacy(pda.other,
+				                             (pda.action == Widelands::DiplomacyAction::kInvite ?
+                                             Widelands::DiplomacyAction::kRefuseInvite :
+                                             Widelands::DiplomacyAction::kRefuseJoin),
+				                             pda.sender);
+			} else {
+				if (planned_action != kNoAction) {
+					verb_log_dbg_time(
+					   gametime,
+					   "AI Diplomacy: Player(%d) replaces plan: old priority: %d, new priority: %d",
+					   static_cast<unsigned int>(mypn), plan_priority, priority);
+					if (planned_action != Widelands::DiplomacyAction::kLeaveTeam) {
+						verb_log_dbg_time(gametime,
+						                  "AI Diplomacy: Player(%d)%s denies the %s of player (%d) with "
+						                  "diploscore %d%s\n",
+						                  static_cast<unsigned int>(mypn), myts_s.c_str(),
+						                  planned_action == Widelands::DiplomacyAction::kAcceptInvite ?
```

Comment:
I believe this will always fall to the second part as we only assign kAcceptInvite further down.

Reply:
Starting at line 3503:
```
				planned_action = pda.action == Widelands::DiplomacyAction::kInvite ?
                                Widelands::DiplomacyAction::kAcceptInvite :
                                Widelands::DiplomacyAction::kAcceptJoin;
```

Reply:
yes I saw that but here in line 3491 we already evaluate for a value that might get assigned only later in the Program run. so here the DiplomacyAction never can be KAcceptInvite can't it?

Reply:
This is inside of the loop over pending diplomacy actions, and `planned_action` is defined outside, and will be finally used outside again. During the first execution it may be `kNoAction` or `kLeaveTeam`, then that may get overridden by `kAcceptInvite` or `kAcceptJoin`. During the later iterations, it could be any of these 4 values.

```
void diplomacy_actions(…) {
   …
   Wl::DiplAction planned_action = kNoAction;
   …
   if (<condition_to_leave>) {
      planned_action = kLeaveTeam;
   }
   …
   for (… pda : game().pending_diplomacy_actions()) {
      <evaluate action>
      if (!accept) {
         <log and execute deny>
      } else {
         if (<previously planned is worse>) {
            <log replacing old action and execute its deny counterpart>
            <line 3491 is here>
         }
         <replace action>
         <now line 3504 is here:> planned_action = <new action>
      }
   }
   <execute planned_action if any>
}
```

Reply:
ok now I get it. Lets have it then